text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+The whole process puts an additional maintenance burden on the project
+maintainers and project resources. This is not the type of work that a casual
+contributor to Paketo will do and as we add more dependencies the burden only
+increases on the maintainer teams.
+
+The motivation of this proposal is to…
+
+- Reduce the burden and toil for Paketo buildpack maintainer teams
+- Continue publishing dependency updates in a timely and secure manner
+- Decouple installing dependencies from configuring them
+- Separate metadata and the actual dependencies so they can be provided to
+  buildpacks in a number of different and flexible ways
+- Establish a reasonable release schedule for buildpacks that’s based around
+  development, not dependencies and thus enabling buildpacks to support version
+  lines
+- Make it easier to package buildpacks for offline environments.
+
+## Detailed Explanation
+At a high level:
+
+- We will define a metadata format that includes the metadata fields currently
+  present in `buildpack.toml`.
+- We will add a dependency version validation section to `buildpack.toml`
+  metadata, this can be used to state that a buildpack version only supports
+  certain ranges of a given tool, such as Java `11.*` or Node.js `16.*`.
+- Metadata can be provided by anyone. A user can add custom metadata, or source
+  from a third party project. Paketo will provide an official set of metadata
+  against which we will test the Paketo buildpacks. It will be distributed via
+  images in an image registry (Docker Hub).
+- Buildpacks do not care how dependency metadata is distributed, that is a
+  separate concern, instead, they just read metadata from a specified location.
+  It is on the platform to mount the metadata that it fetched at that location.
+- The actual dependencies are accessed via the metadata and that can happen
+  over any protocol (HTTPS/SFTP/FILE) and be distributed in any format
+  (archive/image).
+- Dependency metadata will be removed from `buildpack.toml`
+
+
+## Metadata Format
+The metadata presented to a buildpack will be structured as a directory of
+dependencies.
+
+Like it does now, each dependency will have an id but unlike the present
+situation, the id is namespaced. An id is composed of an organization name and
+a dependency name. It follows the [reverse domain name notation](https://en.wikipedia.org/wiki/Reverse_domain_name_notation) and the
+dependency name is defined as the final item, so `com.example.dep-a` would have
+an organization of `com.example` and a dependency of `dep-a`. It is case
+insensitive, so `dep-a` is no different than `Dep-A`. This is to reduce the
+possibility of [typosquatting](https://en.wikipedia.org/wiki/Typosquatting).
+Allowed characters are [the same as for a valid hostname](https://en.wikipedia.org/wiki/Hostname#Syntax).
+
+The directory structure will contain a folder for each dot-separated segment of
+the dependency’s organization name and in the lowest level directory there will
```

Comment:
I would expect that if you provided metadata via multiple images that the container runtime would handle the overlaying of filesystems. I expect that the same would be true when using platform volume mounting, but I'm not familiar enough with the platform spec to know if that's a reasonable assumption.

Reply:
I don't think that overlaying volume mounts is a given, but I definitely share the lack of knowledge to be sure.

This rfc seems to explicitly rely on someone to produce a single source - hence the mentioning of tooling that can take multiple sources and merge it into one. I would think though that this is less dynamic, similar to how one could produce custom stacks and builders - there just quite some effort and automation involved in it.

Reply:
I see that. I'm assuming that for the default case, there will be a single source per Paketo language family. I.e. the Java buildpack maintainers would take the existing dependencies out of the existing buildpacks and move them to a single source that they create and maintain.

Flipping this around, what other system would you propose? We could use guids for dependency layouts, which (effectively) guarantees we won't have filesystem clashes but comes at the cost of readability on the filesystem.

Reply:
One idea could be to introduce a fs node for the dependency sources, like `/platform/dependencies/metadata/io.paketo/org/apache/tomcat`... But that's just a rough thought - will look into this tomorrow.

Reply:
Ok, let us know if you encounter any issues with that exploration.

Regardless, I think this RFC doesn't have to worry about the mechanism by which the filesystem is created and populated - I think it's out of scope. I think we just have to agree on the filesystem layout and who gets ownership of what level in the filesystem hierarchy.

Reply:
@ForestEckhardt this could as well form a compromise for the above. The Paketo Java subteam could deliver a `io.paketo.java` dependency *bundle* that then containes metadata for - amongst others - `org.apache.tomcat`. WDYT?

Although, this might make it more difficult, to define precedence and for example allow a user to override the Tomcat dependency metadata provided by Paketo.

Reply:
@loewenstein I think we're roughly on the same page, but I think the RFC could probably explain this better.

>Like it does now, each dependency will have an id but unlike the present
>situation, the id is namespaced. An id is composed of an organization name and
>a dependency name.

The namespacing I'd thought about was by buildpack or possibly by buildpack team, not by the dependency project's name, so a more realistic example would be:

```
/platform/dependencies/metadata/io
  |
  --- paketo
          |
          --- java
                  |
                  --- apache-tomcat.toml
....
```

The reason I went with top-level directories is that most of the container solutions for injecting files, like volume mounts, assume that there isn't directory overlap (i.e. you can't volume mount over top of something, layer overlays work similarly, in this case you can overlay stuff but then you mask the files in lower layers which is confusing so its best to not do that). Anyway, having the namespace should allow us to easily split up metadata for different buildpack teams or buildpacks and just mount it into different namespaces.

I'm sure there's more we can do with this, but we should try to think MVC here & just ensure that the building blocks are in place so we can do more on top of this structure.